### PR TITLE
Use more compatible flag name

### PR DESCRIPTION
**Pull request by tfausak**
This flag name was changed without comment in 6e1ff781e062b4e01c7da347e7a3f74c60040c2c. It has caused a few problems:

- https://github.com/commercialhaskell/stack/issues/3345
- https://github.com/fpco/stackage/issues/2755
- https://github.com/fpco/stackage/issues/2759
- https://github.com/fpco/stackage/issues/2842
- https://github.com/haskell/cabal/issues/4686
- https://github.com/hvr/cassava/issues/150

Those problems either caused or accelerated some (attempted) changes in Cabal:

- https://github.com/haskell/cabal/pull/4654
- https://github.com/haskell/cabal/pull/4687
- https://github.com/haskell/cabal/pull/4696

Those problems also caused a change in Stack:

- https://github.com/commercialhaskell/stack/pull/3349

In short: Cabal never had any trouble with this. Stack did. The current master version of Stack can handle flags like this, but no released versions of it can. It's not clear when the next version of Stack will be released. In the meantime, no Stack users can use this package. This commit changes the offending flag to something that Stack can handle. By using a flag that Stack can handle, Stack users can once again use this package, and it can return to Stackage. There are no apparent downsides to using a more compatible flag name.

### CODE REVIEW

**Comment by lwm** [comment]
Thank you for taking some time on this one 💯

**Comment by tfausak** [comment]
@hvr what can I do to help move this along?

**Comment by hvr** [comment]
> Stack users can once again use this package

I don't think that's true. Stack users can use this package just fine. LTS 9.6 has https://www.stackage.org/lts-9.6/package/cassava-0.4.5.1 wheras Nightly 2017-09-29 has https://www.stackage.org/nightly-2017-09-29/package/cassava-0.5.0.0

**Comment by tfausak** [comment]
That's... deliberately missing the point. The flag only changed in [version 0.5.1.0](https://github.com/hvr/cassava/blob/v0.5.0.0/cassava.cabal#L25), so of course Stack users can install versions before that. Stack users cannot install the latest version of this package because the flag name is invalid for all released versions of Stack.

``` sh
> stack --version
Version 1.5.1 x86_64 hpack-0.17.1

> stack --resolver lts-9.0 build cassava-0.5.1.0
Error: While constructing the build plan, the following exceptions were encountered:
In the dependencies for cassava-0.5.1.0:
    bytestring-0.10.8.1 must match >=0.9.2 && <0.10.4 (latest applicable is 0.10.2.0)
needed for unknown reason - stack invariant violated.
Plan construction failed.

> stack --resolver lts-9.0 build --flag cassava:-bytestring--LT-0_10_4 cassava-0.5.1.0
option --flag: Invalid flag name: bytestring--LT-0_10_4
```

Why is it important that the flag name be like this?

**Comment by hvr** [comment]
Let me turn the burden of proof around, and ask why does it take so long for Stack to make a point release even though this bug in Stack has been known for over a month already, when I made it quite clear in https://github.com/commercialhaskell/stack/issues/3345#issuecomment-322395708 that Stack's the place where this ought to be fixed. So I'd suggest you try shaming Stack dev into releasing a point release instead.

**Comment by tfausak** [comment]
I agree that it would be nice for Stack to cut a release that supports these flag names. However I think it's clear that it's easier to change a flag name in Cassava than it is to release a new version of Stack. 

Regardless, that doesn't explain why this flag name has to be the way it is. What was wrong with the old flag (or my proposed flag)?

**Comment by tfausak** [comment]
For anyone trying to install `cassava-0.5.1.0` with Stack 1.5.1, I released my fork on Hackage as [`Cassava-0.5.1.0`](https://hackage.haskell.org/package/Cassava-0.5.1.0). You can install it like so:

``` sh
> stack --resolver lts-9.0 build --flag Cassava:-bytestring-LT-0_10_4 Cassava-0.5.1.0 text-short-0.1.1
```

Other than the flag name, it's identical to `cassava-0.5.1.0`.

**Comment by mgsloan** [comment]
@hvr Even if stack fixes this, older versions of stack will still not work with this package.  Not all distributions and users keep up to date.

Your criticism of stack's release cycle seems hilarious considering cabal-install's history of releases.  Even so, I'm not sure we will "fix" this at all.  IMHO there should be a single set of rules for package names and flag names.  Less to remember.  So, IMHO, this is an oversight in cabal's spec and implementation.  In order to make it so that most current build configurations will build with prior versions of stack, I think it makes sense to keep the restriction in stack.

Ideally, to maintain compatibility between the tools, cabal should add the restriction.  I doubt this will break many existing packages, because it is such a strange corner case.

I have seen you give no justification for this flag name.  It seems that its purpose is solely to break compilation with stack.  While at times the relationship is tenuous, I think overall there has been an amicable relationship between hackage / cabal-install and stackage / stack.   From our perspective, this is a petty and childish action on your part which partially compromises future amicable collaboration.  We will gladly continue collaborating with community members who are more reasonable.

Please reconsider your decision to flagrantly disregard the stack use-case.  It is not too late to restore good will.

I realize that you may draw an analogy to many packages not having upper bounds.  However, many intelligent people, perhaps most, will agree there are good reasons to omit upper bounds.  On the other hand, there are no good reasons to have a double hyphen in a flag name.  The only purpose here is to break usage of this package with stack.

**Comment by mgsloan** [comment]
After some discussion with others via email and slack, the following is now clear:  @hvr has a tool that generates these flag names, and likely did not initially realize that they would cause problems for stack

So, my attribution of definitive malice to this choice seems to be inaccurate.  Sorry for tossing around accusations.  With the comments here, HVR was making it quite hard to not attribute malice.  From the perspective of a stack developer, stubbornness to just change an inconsequential name seems quite malicious.  At the very least, not very friendly.

FWIW, typical stack configurations are currently quite forwards and backwards compatible.  The very first released version of stack can still build the stack of a few months ago, and vice versa.  It probably works the other way too.  While this certainly isn't guaranteed, it really sucks to have some packages unnecessarily break this property.

I agree that this is a bug in stack.  However, by not working around this bug, you are implicitly saying that you don't care about the users of these packages.  Or at least, not the ones that use older versions of stack.

**Comment by tonymorris** [comment]
Programmers :)

**Comment by mitchellwrosen** [comment]
Man... was releasing `Cassava-0.5.1.0` (note: capital C) really the right move here? Heh. If you had to grab this package in a pinch I'm sure a non-Hackage dependency would have sufficed until you could get this issue sorted out. But now we have two different `[cC]assava` packages forever. Confused emoji face.

**Comment by mitchellwrosen** [comment]
@tfausak says via email, posted w/ his permission:

> If `cassava` fixes the flag name, I'll gladly deprecate `Cassava` in favor of it. It'll be just like the [`cabal`](https://hackage.haskell.org/package/cabal) package.

**Comment by Cipherwraith** [comment]
Its unfortunate that there is an impasse of such massive proportions between the stack devs and @hvr.

**Comment by ekmett** [comment]
Dumb question: 

Would changing the flag actually fix anything if stack has to parse the old files anyways? 

It seems based on cursory observation that once _any_ version of any package with a flag name that stack didn't understand made its way into the index, it seems stack is just hosed, no?

I'd be happy to be wrong here.

**Comment by ekmett** [comment]
https://twitter.com/taylorfausak/status/948280024129527808 makes a fairly compelling case for renaming the flag in my eyes. It'd mean even old versions of stack could build new snapshots.

**Comment by ejmg** [comment]
wasted nearly 45 mins diagnosing this bug because of a divergence between my shell instances and upgrading stack, silly this PR wasn't taken for something so simple yet time draining. Thanks!

**Comment by tonymorris** [comment]
I didn't hit this issue and my team spent 0 minutes on it. That's because
we don't use stack.

**Comment by ejmg** [comment]
Hey, thanks for the feedback. Tons of people use Stack because it's a pretty great tool, popular trade books endorse it, and large companies like Facebook actively advocate its adoption, so I'm sure your feedback will really turn the tide here. 

To everyone else I'm spamming on this thread, I'm sorry [my comment](https://github.com/haskell-hvr/cassava/pull/155#issuecomment-340624727) didn't more appropriately join the community discussion on whether certain flag naming choices in a major library should accommodate a major build tool. It was my first project with the language and I got frustrated after the lost time from a small bug/realizing my code was actually fine.

**Comment by tonymorris** [comment]
Yeah, I remember in the late 1990s when a large company endorsed that we all use Windows 98 and plenty of people were using it. Turns out that was a bad idea!

Sorry to hear your introduction was frustrating. Try not using stack and see how that goes. We're all waiting to help you.
